sw/device/lib/crypto/drivers/entropy.c

```diff
@@ -0,0 +1,183 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/crypto/drivers/entropy.h"
+
+#include "sw/device/lib/base/abs_mmio.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+
+#include "csrng_regs.h"  // Generated
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+enum {
+  kBaseCsrng = TOP_EARLGREY_CSRNG_BASE_ADDR,
+
+  /**
+   * CSRNG genbits buffer size in uint32_t words.
+   */
+  kEntropyCsrngBitsBufferSize = 4,
+};
+
+/**
+ * Supported CSRNG application commands.
+ * See https://docs.opentitan.org/hw/ip/csrng/doc/#command-header for
+ * details.
+ */
+typedef enum entropy_csrng_op {
+  kEntropyDrbgOpInstantiate = 1,
+  kEntropyDrbgOpReseed = 2,
+  kEntropyDrbgOpGenerate = 3,
+  kEntropyDrbgOpUpdate = 4,
+  kEntropyDrbgOpUnisntantiate = 5,
+} entropy_csrng_op_t;
+
+/**
+ * CSRNG application interface command header parameters.
+ */
+typedef struct entropy_csrng_cmd {
+  /**
+   * Application command ID.
+   */
+  entropy_csrng_op_t id;
+  /**
+   * Entropy source enable.
+   *
+   * Mapped to flag0 in the hardware command interface.
+   */
+  hardened_bool_t disable_trng_input;
+  const entropy_seed_material_t *seed_material;
+  /**
+   * Generate length. Specified as number of 128bit blocks.
+   */
+  uint32_t generate_len;
+} entropy_csrng_cmd_t;
+
+OT_WARN_UNUSED_RESULT
+static status_t csrng_send_app_cmd(uint32_t reg_address,
+                                   entropy_csrng_cmd_t cmd) {
+  uint32_t reg;
+  bool cmd_ready;
+  bool cmd_error;
+  do {
+    reg = abs_mmio_read32(kBaseCsrng + CSRNG_SW_CMD_STS_REG_OFFSET);
+    cmd_ready = bitfield_bit32_read(reg, CSRNG_SW_CMD_STS_CMD_RDY_BIT);
+  } while (!cmd_ready);
+
+  // The application command header is not specified as a register in the
+  // hardware specification, so the fields are mapped here by hand. The
+  // command register also accepts arbitrary 32bit data.
+  static const uint32_t kAppCmdBitFlag0 = 8;
+  static const bitfield_field32_t kAppCmdFieldCmdId = {.mask = 0xf, .index = 0};
+  static const bitfield_field32_t kAppCmdFieldCmdLen = {.mask = 0xf,
+                                                        .index = 4};
+  static const bitfield_field32_t kAppCmdFieldGlen = {.mask = 0x7ffff,
+                                                      .index = 12};
+
+  uint32_t cmd_len = cmd.seed_material == NULL ? 0 : cmd.seed_material->len;
+
+  if (cmd_len & ~kAppCmdFieldCmdLen.mask) {
+    return INTERNAL();
+  }
+
+  // TODO: Consider removing this since the driver will be constructing these
+  // commands internally.
+  // Ensure the `seed_material` array is word-aligned, so it can be loaded to a
+  // CPU register with natively aligned loads.
+  if (cmd.seed_material != NULL &&
+      misalignment32_of((uintptr_t)cmd.seed_material->data) != 0) {
+    return INTERNAL();
+  }
```

Comment:
I agree with the `TODO`. nit: may want to consider adding an assertion of the alignment of this field, in case the data type gets changed to something else, e.g. `char`, in the future.